lib/src/conflicts.rs

```diff
@@ -28,11 +29,29 @@ use crate::merge::{Merge, MergeBuilder, MergedTreeValue};
 use crate::repo_path::RepoPath;
 use crate::store::Store;
 
-const CONFLICT_START_LINE: &[u8] = b"<<<<<<<\n";
-const CONFLICT_END_LINE: &[u8] = b">>>>>>>\n";
-const CONFLICT_DIFF_LINE: &[u8] = b"%%%%%%%\n";
-const CONFLICT_MINUS_LINE: &[u8] = b"-------\n";
-const CONFLICT_PLUS_LINE: &[u8] = b"+++++++\n";
+const CONFLICT_START_LINE: &[u8] = b"<<<<<<<";
+const CONFLICT_END_LINE: &[u8] = b">>>>>>>";
+const CONFLICT_DIFF_LINE: &[u8] = b"%%%%%%%";
+const CONFLICT_MINUS_LINE: &[u8] = b"-------";
+const CONFLICT_PLUS_LINE: &[u8] = b"+++++++";
+const CONFLICT_START_LINE_CHAR: u8 = CONFLICT_START_LINE[0];
+const CONFLICT_END_LINE_CHAR: u8 = CONFLICT_END_LINE[0];
+const CONFLICT_DIFF_LINE_CHAR: u8 = CONFLICT_DIFF_LINE[0];
+const CONFLICT_MINUS_LINE_CHAR: u8 = CONFLICT_MINUS_LINE[0];
+const CONFLICT_PLUS_LINE_CHAR: u8 = CONFLICT_PLUS_LINE[0];
+
+/// A conflict marker is one of the separators, optionally followed by a space
+/// and some text.
+// TODO: All the `{7}` could be replaced with `{7,}` to allow longer
+// separators. This could be useful to make it possible to allow conflict
+// markers inside the text of the conflicts.
+static CONFLICT_MARKER_REGEX: once_cell::sync::Lazy<Regex> = once_cell::sync::Lazy::new(|| {
+    Regex::new(
+        r"(<{7}|>{7}|%{7}|\-{7}|\+{7})( .*)?
+",
+    )
+    .unwrap()
+});
```

Comment:
Maybe this is nitpicky but I don't see why a regex is needed here at all when it's only used as a shorthand to match against a handful of possible outcomes? We could just write a 10 line string handling routine here, right?

Reply:
What do you have in mind? I find a regex to be a good way to write a string handling routine.

Not sure whether these are helpful, but:

- I was initially going to allow separators of length more than 7, and this might happen later.
- I want to allow only a newline or a space right after the separator
- I'm not sure whether at some point we will want to allow CRLF here.
- I was considering using a https://docs.rs/regex/latest/regex/struct.RegexSet.html, but I don't think there's a need so far.

Reply:
TBH it's mostly just a pet peeve, I guess. In this case all we're really doing is taking the length-7 substring of the line and seeing if it matches any of 5 different strings, right? e.g. `True if s[0:6] in ["...", "..."] else False` in something like Python. I don't have numbers to back this up but I'd also suspect it helps code size and efficiency since bringing in all the regex machinery requires parsing the input, building the state machine, etc, not to mention the cost of inlining the code into the call site and optimizing etc. I wouldn't block this or anything over it though, of course. It just seems like a slightly larger hammer than necessary I guess.

Reply:
I usually assume that the regex engine will come up with some sort of state machine that will be at least as efficient as naive code.

I think Rust uses a very fast library with guarantees on computational complexity (for example, it doesn't support backtracking; I'm also not a regex nerd so I'm not entirely sure, but https://docs.rs/regex/latest/regex/index.html supports my statement), likely written by the author of `ripgrep`. It's a much smaller hammer than a full possibly-exponential Perl regex engine.

Again, I'd like it to be easy to allow longer-than-7 separators later. Already, it'd be a bit more complex than what you suggested: `s[0:6] in [...] and s[7] in [" ", "\n"]`.

Actually, I think my main problem is that I don't immediately know how to write that elegantly in Rust. If you'd like to suggest something, feel free and I'll see if I can adapt it. I certainly don't find my code perfectly elegant.

> all the regex machinery requires parsing the input, building the state machine

I made sure to do this once globally.

Reply:
I'll keep it as is for now, but we can clean things up in a follow-up.